src/thermo/BinarySolutionTabulatedThermo.cpp

```diff
@@ -0,0 +1,230 @@
+        } else if (xnow == 1)
+        {
+            dS_corr = BigNumber;
+        } else
+        {
+            dS_corr = GasConstant*std::log(xnow/(1.0-xnow)) + GasConstant/Faraday*std::log(this->standardConcentration(this->nSpecies()-m_kk_tab)/this->standardConcentration(m_kk_tab));
+        }
+        c[2] = d[1] * 1e3 + dS_corr; // 1e3 for conversion J/K/mol -> J/K/kmol
+        c[3] = 0.0;
+        type = m_spthermo.reportType(m_kk_tab);
+        tlow = m_spthermo.minTemp(m_kk_tab);
+        thigh = m_spthermo.maxTemp(m_kk_tab);
+        shared_ptr<SpeciesThermoInterpType> stit(
+                newSpeciesThermoInterpType(type, tlow, thigh, OneAtm, c));
+        m_spthermo.modifySpecies(m_kk_tab, stit);
+        // Update the thermodynamic functions of the reference state.
+        m_spthermo.update(tnow, m_cp0_R.data(), m_h0_RT.data(), m_s0_R.data());
+        doublereal rrt = 1.0 / RT();
+        for (size_t k = 0; k < m_kk; k++) {
+            double deltaE = rrt * m_pe[k];
+            m_h0_RT[k] += deltaE;
+            m_g0_RT[k] = m_h0_RT[k] - m_s0_R[k];
+        }
+        m_xlast = xnow;
+        m_tlast = tnow;
+    }
+}
+
+void BinarySolutionTabulatedThermo::initThermoXML(XML_Node& phaseNode, const std::string& id_)
+{
+    vector_fp x, h, s;
+    std::vector<std::pair<double,double>> x_h_temp, x_s_temp;
+
+    if (id_.size() > 0) {
+        if (phaseNode.id() != id_) {
+            throw CanteraError("BinarySolutionTabulatedThermo::initThermoXML",
+                    "phasenode and Id are incompatible");
+        }
+    }
+    if (nSpecies()!=2) {
+    	throw CanteraError("BinarySolutionTabulatedThermo::initThermoXML",
+    	                    "No. of species should be equal to 2!");
+    }
+    if (phaseNode.hasChild("thermo")) {
+        XML_Node& thermoNode = phaseNode.child("thermo");
+        std::string mString = thermoNode["model"];
+        if (!caseInsensitiveEquals(mString, "binarysolutiontabulatedthermo")) {
+            throw CanteraError("BinarySolutionTabulatedThermo::initThermoXML",
+                    "Unknown thermo model: " + mString);
+        }
+        if (thermoNode.hasChild("tabulatedSpecies")) {
+        	XML_Node& speciesNode = thermoNode.child("tabulatedSpecies");
+            std::string tabulated_species_name = speciesNode["name"];
+            m_kk_tab = speciesIndex(tabulated_species_name);
+            if (m_kk_tab == npos) {
+                throw CanteraError("BinarySolutionTabulatedThermo::initThermoXML",
+                        "Species " + tabulated_species_name + " not found.");
+            }
+            m_xlast = moleFraction(m_kk_tab);
+        }
+        if (thermoNode.hasChild("tabulatedThermo")) {
+        	XML_Node& dataNode = thermoNode.child("tabulatedThermo");
+            getFloatArray(dataNode, x, false, "", "moleFraction");
+            getFloatArray(dataNode, h, false, "", "enthalpy");
+            getFloatArray(dataNode, s, false, "", "entropy");

```

Comment:
These arrays should be converted to Cantera's internal units here (i.e. SI+kmol). As written, the units specified in the CTI file are being ignored.

Reply:
Hmm, even worse: they're assumed to be J/mol and J/mol/K (see lines 56 and 67 in BinarySolutionTabulatedThermo.cpp).  So ignored and assumed of a form different from Cantera's internal units.  Will fix in a second.

Reply:
Do I leave the moleFraction entry as is, since it is unitless/dimensionless?  Or is there some entry which specifies this?